### PR TITLE
[ml/train] `Trainer` [3/4]: `TorchTrainer` implementation

### PR DESCRIPTION
**Pull request by amogkam**
Initial implementation for training on Ray AIR part 3.

Adds `TorchTrainer` to Ray AIR.

Depends on
https://github.com/ray-project/ray/pull/22921, https://github.com/ray-project/ray/pull/22929

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/integrations/torch.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+from ray.train.torch import TorchConfig
+from ray.train.ml_trainer import GenDataset
+from ray.train.impl.data_parallel_trainer import DataParallelFunctionTrainer
+from ray.ml.config import ScalingConfig, RunConfig
+from ray.ml.preprocessor import Preprocessor
+from ray.ml.checkpoint import Checkpoint
+
+
+class TorchTrainer(DataParallelFunctionTrainer):
+    """
```
Comment (ericl) [inline python/ray/train/integrations/torch.py:12]: Remove newline after the """

File: python/ray/train/integrations/torch.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    A Trainer for data parallel PyTorch training.
+
+    This Trainer runs the provided function on multiple Ray Actors. The
+    provided datasets are automatically sharded to each Ray actor running
+    the training function.
+
+    Args:
+        train_func (Callable): The training function to execute.
```
Comment (ericl) [inline python/ray/train/integrations/torch.py:20]:
```suggestion
        train_func: The training function to execute.
```

File: python/ray/train/integrations/torch.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        torch_config (Optional[TorchConfig]): Configuration
+            for setting up the PyTorch backend. If set to None, use the
+            default configuration.
+        scaling_config (Optional[ScalingConfig]): Configuration
+            for how to scale data parallel training.
+        run_config (Optional[RunConfig]): Configuration for the execution of
+            the training run.
+        train_dataset (Optional[GenDataset]): Either a distributed Ray
+            :ref:`Dataset <dataset-api>` or :ref:`DatasetPipeline
+            <dataset-pipeline-api>`, or a Callbable that returns a Dataset, to
```
Comment (ericl) [inline python/ray/train/integrations/torch.py:33]:
```suggestion
            <dataset-pipeline-api>`, or a Callable that returns a Dataset to
```

File: python/ray/train/impl/data_parallel_trainer.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
+import logging
+import os
+from typing import Dict, Callable, Optional, Type, TYPE_CHECKING
+
+import ray
+from ray.ml.preprocessor import Preprocessor
+from ray.ml.checkpoint import Checkpoint
+from ray.ml.config import ScalingConfig, RunConfig, _DataParallelScalingConfig
```
Comment (ericl) [inline python/ray/train/impl/data_parallel_trainer.py:8]: Should probably make DataParallelScalingConfig public if most trainers are going to use it.

File: python/ray/train/integrations/torch.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            train_func=train_func,
+            train_func_config=train_func_config,
+            backend_config=torch_config,
+            scaling_config=scaling_config,
+            run_config=run_config,
+            train_dataset=train_dataset,
+            additional_datasets=additional_datasets,
+            preprocessor=preprocessor,
+            resume_from_checkpoint=resume_from_checkpoint,
+        )
```
Comment (ericl) [inline python/ray/train/integrations/torch.py:72]: Huh, is the actual implementation of the TorchTrainer left for another PR? I expected there would be some torch specific code here.

Comment (amogkam) [reply python/ray/train/integrations/torch.py:72]: No, this is all handled by the superclass and Ray Train.

**Pull request closed without merging**